File: modules/auth_service/src/session.cpp
#include "oink_judge/auth_service/session.h"

#include <ctime>
#include <oink_judge/config/config.h>
#include <oink_judge/utils/crypto.h>
#include <random>

namespace oink_judge::auth_service {

namespace {

std::random_device rd;     // NOLINT
std::mt19937_64 rng(rd()); // NOLINT

} // namespace

using Config = config::Config;

Session::Session(std::string username) : username_(std::move(username)), expire_at_(0) {}

auto Session::generateSession() -> void {
    if (!session_id_.empty()) {
        return;
    }

    session_id_ = generateSessionId();
    expire_at_ = generateExpiredAt();
}

auto Session::isValid() const -> bool {
    time_t current_time = std::time(nullptr);
    return current_time <= expire_at_;
}

auto Session::getSessionId() const -> std::string {
    if (!isValid()) {
        return "";
    }

    return session_id_;
}

auto Session::getUsername() const -> const std::string& { return username_; }

auto Session::getExpireAt() const -> time_t { return expire_at_; }

auto Session::generateSessionId() -> std::string {
    time_t current_time = std::time(nullptr);
    std::string data = std::to_string(current_time) + "." + std::to_string(rng()) + "." + getUsername();

    return utils::crypto::sha256(data);
}

auto Session::generateExpiredAt() -> time_t {
    return std::time(nullptr) + static_cast<time_t>(Config::config()["bounds"]["valid_session_time"]);
}

} // namespace oink_judge::auth_service

File: app/services/auth/session.py
import hashlib
import secrets
import time

from app.config import get_timing, require_has_value


class Session:
    def __init__(self, username: str) -> None:
        self._username = username
        self._session_id = ""
        self._expire_at = 0

    def generate_session(self) -> None:
        if self._session_id:
            return
        self._session_id = self._generate_session_id()
        self._expire_at = self._generate_expired_at()

    @property
    def session_id(self) -> str:
        if not self.is_valid:
            return ""
        return self._session_id

    @property
    def username(self) -> str:
        return self._username

    @property
    def expire_at(self) -> int:
        return self._expire_at

    @property
    def is_valid(self) -> bool:
        return time.time() <= self._expire_at

    def _generate_session_id(self) -> str:
        data = f"{int(time.time())}.{secrets.randbelow(2**64)}.{self._username}"
        return hashlib.sha256(data.encode()).hexdigest()

    def _generate_expired_at(self) -> int:
        timing = get_timing("valid_session_time")
        if timing is None:
            seconds = 3600
        else:
            seconds = int(require_has_value(timing).total_seconds())
        return int(time.time()) + seconds
